src/libraries/System.Text.Json/docs/KeyConverter_spec.md

```diff
@@ -0,0 +1,185 @@
+# Extend support for TKey in Dictionaries to non-string types
+
+# Motivation
+Most of our users that serialize dictionary use `Dictionary<string, TKey>`; however there is a significant amount that relies on `Dictionary<TKey, TValue>` where `TKey` is a primitive other than `string`, e.g: `int` or `Guid`, most of them came form `Newtonsoft.Json` which offers a plenty amount of support for using several types as the `TKey`, other popular .Net serializers also offer support for other types, the most common are integers (`int`, `uint`, `long`, etc.), and `enum`s (including Flags `enum`s).
+
+# Goals
+* 80%+ of dictionaries with non-string keys work out of the box, especially if they can round-trip.
+* Remain high performance.
+
+# Non-goals 
+* Complete parity with `Newtonsoft.Json` capabilities, especially in how string support is extended; any extension point can be through `JsonConverter<MyDictionary<non-string, TValue>>`.
+
+# Sample
+```cs
+// (De)serialize into a dictionary with a non-string key.
+Dictionary<int, string> root = new Dictionary<int, string>();
+root.Add(1, "value");
+
+string json = JsonSerializer.Serialize(root);
+// JSON
+// {
+//   "1":"value"
+// }
+
+Dictionary<int, string> rootCopy = JsonSerializer.Deserialize<Dictionary<int, string>>(json);
+Console.WriteLine(rootCopy[1]);
+ // Prints 
+ // value
+```
+
+# Strawman proposal
+
+## `KeyConverter`
+Implement an internal custom mechanism that is in charge of converting a defined set of types to be supported as the dictionary `TKey`; more or less like internal `JsonConverter`s work but for dictionary keys to JSON property names and viceversa.
+
+* The alternative that offers the best performance.
+
+* We need to define a criteria to choose what types we should support, I suggest to do as Utf8JsonReader/Writer and support the types supported by the Utf8Parser/Formatter.
+
+* Supported types (Types supported by [`Utf8Formatter/Parser`](https://docs.microsoft.com/en-us/dotnet/api/system.buffers.text.utf8formatter?view=netcore-3.1) + a few others that are popular):
+  * Boolean
+  * Byte
+  * DateTime
+  * DateTimeOffset
+  * Decimal
+  * Double
+  * Enum
+  * Guid
+  * Int16
+  * Int32
+  * Int64
+  * Object (Only on Serialization and if the runtime type is one of the supported types in this list)
+  * SByte
+  * Single
+  * String
+  * UInt16
+  * UInt32
+  * UInt64
```

Comment:
Add `char`?

Reply:
Sure, I will add it. 

I missed that one since the `Utf8Formatter` does not have a signature for it but the `JsonSerializer` supports it just fine.

Reply:
What do you think about adding `byte[]` to this list?

Reply:
@am11 We can support it and do as we do for a `byte[]` value, which is write/read it as a Base64 string.

Reply:
For floating point (Single and Double) does GetHashCode() and Equals() work as expected? Since a given Single value may be 1.23 for example but can be represented in more than one binary representation (i.e. the 4 bytes may be different). By default Equals() compare bytes which wouldn't work correctly in all cases @tannergooding

Reply:
For float/double (and IEEE binary floating-point numbers in general), every bit representation for finite, non-zero floating-point values represents something unique.

The exceptions are:
* +0.0 and -0.0
* The many representations of NaN

Both `==`/`!=` and `Equals` then perform standard equality comparisons for all other values (all non-zero finite values, and +/-infinity).
For `==`/`!=` and `Equals`, `+0` == `-0`

For `==`/`!=`, any comparison where one operand is `NaN` returns `false`
For `Equals`, `NaN` is equal to any other `NaN`, regardless of underlying bit representation.

`GetHashCode` follows the same rule as `Equals`. That is `+0` and `-0` return the same hash code. All `NaN` return the same hash code.

Reply:
OK thanks. So using Single\Double as a dictionary key makes sense. After looking at the binary representation for the non-exception cases I see the exponent is always normalized by selecting the smallest value that produces the representation (so binary representation of 1.23 == 123e-2 for example).